designdocs/message_storage_v2.md

```diff
@@ -0,0 +1,68 @@
+# On Chain Message Storage
+
+## Context and Scope
+The proposed feature consists of changes that is going to be one (or more) pallet(s) in runtime of a
+Substrate based blockchain, and it will be used in all environments including production.
+
+## Problem Statement
+After introduction of **Proof of Validity** or **PoV** in runtime weights, all pallets should be
+re-evaluated and refactored if necessary to minimize the usage of **PoV**. This is to ensure all
+important operations are scalable.
+This document tries to propose some changes on **Messages** pallet to optimize the **PoV** size.
+
+## Goals
+- Minimizing Weights including **execution times** and **PoV** size.
+
+## Proposal
+Storing messages on chain using **BlockNumber** and **SchemaId** and **Index** as main and secondary
+ and tertiary keys using [StorageNMap](https://paritytech.github.io/substrate/master/frame_support/storage/trait.StorageNMap.html) data structure provided in Substrate.
+
+### Main Storage types
+- **MessagesV2**
+    - _Type_: `StorageNMap<(BlockNumber, SchemaId, u16), Message>`
+    - _Purpose_: Main structure To store all messages for a certain block number and schema id and
+      index
+- **BlockMetadata**
+    -  _Type_: `StorageValue<_, BlockMetadata`
+    - _Purpose_: Stores temporary (Whitelisted) metadata for that block
```

Comment:
nit: term  `BlockMeta' maybe too generic ? It's pretty much message attributes

Reply:
It is but it's related to the current bock. maybe `BlockMessageMetadata` ? Open to other suggestions